File: BTFViewer/btf_viewer_pkg/btf_slice.py
"""Export a cursor-range slice of a BTF text stream (Desktop + CLI)."""
from __future__ import annotations

import gzip
import os
from typing import Iterable, List, Tuple

from .parser import BtfTrace, _open_btf_text, _task_display_name


def filter_btf_text_to_range(text: str, lo: int, hi: int) -> Tuple[str, int]:
    """Keep ``#`` meta, ``C`` (set_frequency) lines, and events in ``[lo, hi]``."""
    if hi < lo:
        lo, hi = hi, lo
    lines, n = _filter_btf_lines(str(text or "").splitlines(), lo, hi)
    return "\n".join(lines) + ("\n" if lines else ""), n


def filter_btf_file_to_range(filepath: str, lo: int, hi: int) -> Tuple[str, int]:
    """Re-read *filepath* and keep events whose timestamp is in ``[lo, hi]``."""
    if hi < lo:
        lo, hi = hi, lo
    with _open_btf_text(filepath) as fh:
        lines, n = _filter_btf_lines(fh, lo, hi)
    return "\n".join(lines) + ("\n" if lines else ""), n


def write_btf_text(text: str, dest_path: str) -> None:
    """Write UTF-8 BTF; ``.gz`` destinations are gzip-compressed."""
    dest = os.path.abspath(dest_path)
    data = str(text or "").encode("utf-8")
    parent = os.path.dirname(dest)
    if parent:
        os.makedirs(parent, exist_ok=True)
    low = dest.lower()
    if low.endswith(".gz"):
        with gzip.open(dest, "wb") as fh:
            fh.write(data)
        return
    with open(dest, "wb") as fh:
        fh.write(data)


def reconstruct_btf_slice(trace: BtfTrace, lo: int, hi: int) -> Tuple[str, int]:
    """Build a best-effort BTF when the original source text is unavailable."""
    if hi < lo:
        lo, hi = hi, lo
    lines: List[str] = []
    for key, val in dict(getattr(trace, "meta", None) or {}).items():
        lines.append(f"#{key} {val}")
    scale = getattr(trace, "time_scale", "") or ""
    if scale and "timeScale" not in (getattr(trace, "meta", None) or {}):
        lines.append(f"#timeScale {scale}")
    lines.append(f"#sliced {lo}-{hi}")
    events: List[Tuple[int, str]] = []
    repr_map = getattr(trace, "task_repr", None) or {}
    for seg in getattr(trace, "segments", None) or []:
        start = int(seg.start)
        end = int(seg.end)
        if end < lo or start > hi:
            continue
        a = max(start, lo)
        b = min(end, hi)
        if a >= b:
            continue
        core = str(seg.core or "Core_0")
        raw = repr_map.get(seg.task, seg.task)
        name = _task_display_name(raw) if raw else str(seg.task)
        name = str(name or seg.task).replace(",", " ")
        events.append((a, f"{a},{core},0,T,{name},0,resume,"))
        events.append((b, f"{b},{core},0,T,{name},0,preempt,"))
    for ev in getattr(trace, "sti_events", None) or []:
        t = int(ev.time)
        if t < lo or t > hi:
            continue
        note = str(ev.note or "").replace("\n", " ")
        events.append((
            t,
            f"{t},{ev.core or 'Core_0'},0,STI,{ev.target},0,{ev.event},{note}",
        ))
    for t in getattr(trace, "tick_sti_times", None) or []:
        ts = int(t)
        if lo <= ts <= hi:
            events.append((ts, f"{ts},Core_0,0,STI,TICK,0,trigger,"))
    events.sort(key=lambda row: (row[0], row[1]))
    for _t, line in events:
        lines.append(line)
    return "\n".join(lines) + ("\n" if lines else ""), len(events)


def _filter_btf_lines(src: Iterable[str], lo: int, hi: int) -> Tuple[List[str], int]:
    out: List[str] = [f"#sliced {lo}-{hi}"]
    kept = 0
    for raw in src:
        line = str(raw).rstrip("\r\n")
        stripped = line.strip()
        if not stripped:
            continue
        if stripped[0] == "#":
            out.append(stripped)
            continue
        parts = stripped.split(",", 8)
        if len(parts) < 4:
            continue
        ev_type = parts[3].strip()
        if ev_type == "C":
            out.append(stripped)
            continue
        try:
            t = int(parts[0].strip())
        except ValueError:
            continue
        if lo <= t <= hi:
            out.append(stripped)
            kept += 1
    return out, kept

File: BTFViewer/web/src/utils/btfSlice.js
/**
 * Cursor-range BTF slice (parity with btf_viewer_pkg/btf_slice.py).
 */

import { taskLabelForMergeKey, taskReprGet } from './colors.js'

export function filterBtfTextToRange(text, lo, hi) {
  let a = Number(lo)
  let b = Number(hi)
  if (!Number.isFinite(a) || !Number.isFinite(b)) {
    throw new Error('Slice range must be numeric timestamps')
  }
  a = Math.trunc(a)
  b = Math.trunc(b)
  if (b < a) [a, b] = [b, a]
  const { lines, kept } = filterBtfLines(String(text || '').split(/\r?\n/), a, b)
  return { text: lines.join('\n') + (lines.length ? '\n' : ''), kept }
}

export function reconstructBtfSlice(trace, lo, hi) {
  let a = Number(lo)
  let b = Number(hi)
  if (!Number.isFinite(a) || !Number.isFinite(b)) {
    throw new Error('Slice range must be numeric timestamps')
  }
  a = Math.trunc(a)
  b = Math.trunc(b)
  if (b < a) [a, b] = [b, a]
  const lines = []
  const meta = trace?.meta && typeof trace.meta === 'object' ? trace.meta : {}
  for (const [key, val] of Object.entries(meta)) {
    lines.push(`#${key} ${val}`)
  }
  if (trace?.timeScale && !meta.timeScale) lines.push(`#timeScale ${trace.timeScale}`)
  lines.push(`#sliced ${a}-${b}`)
  const events = []
  for (const seg of trace?.segments || []) {
    const start = Number(seg.start)
    const end = Number(seg.end)
    if (!Number.isFinite(start) || !Number.isFinite(end)) continue
    if (end < a || start > b) continue
    const s = Math.trunc(Math.max(start, a))
    const e = Math.trunc(Math.min(end, b))
    if (s >= e) continue
    const core = seg.core || 'Core_0'
    const mk = seg.mergeKey || seg.task || ''
    const name = String(
      taskLabelForMergeKey(trace, mk) || taskReprGet(trace, mk) || mk || 'task',
    ).replace(/,/g, ' ')
    events.push([s, `${s},${core},0,T,${name},0,resume,`])
    events.push([e, `${e},${core},0,T,${name},0,preempt,`])
  }
  for (const ev of trace?.stiEvents || []) {
    const t = Math.trunc(Number(ev.time))
    if (!Number.isFinite(t) || t < a || t > b) continue
    const note = String(ev.note || '').replace(/\n/g, ' ')
    events.push([
      t,
      `${t},${ev.core || 'Core_0'},0,STI,${ev.target || ''},0,${ev.event || ''},${note}`,
    ])
  }
  for (const tick of trace?.tickStiTimes || []) {
    const t = Math.trunc(Number(tick))
    if (Number.isFinite(t) && t >= a && t <= b) {
      events.push([t, `${t},Core_0,0,STI,TICK,0,trigger,`])
    }
  }
  events.sort((x, y) => x[0] - y[0] || String(x[1]).localeCompare(String(y[1])))
  for (const [, line] of events) lines.push(line)
  return { text: lines.join('\n') + (lines.length ? '\n' : ''), kept: events.length }
}

function filterBtfLines(src, lo, hi) {
  const out = [`#sliced ${lo}-${hi}`]
  let kept = 0
  for (const raw of src) {
    const stripped = String(raw || '').trim()
    if (!stripped) continue
    if (stripped[0] === '#') {
      out.push(stripped)
      continue
    }
    const parts = stripped.split(',', 9)
    if (parts.length < 4) continue
    const evType = String(parts[3] || '').trim()
    if (evType === 'C') {
      out.push(stripped)
      continue
    }
    const t = Number.parseInt(String(parts[0] || '').trim(), 10)
    if (!Number.isFinite(t)) continue
    if (t >= lo && t <= hi) {
      out.push(stripped)
      kept += 1
    }
  }
  return { lines: out, kept }
}
